pkg/codegen/nodejs/gen.go

```diff
@@ -1617,6 +1635,133 @@ type namespace struct {
+		len(ns.types),
+		len(ns.enums),
+		childrenStr,
+	)
+}
+
+// ioContext defines a set of parameters used when generating input/output
+// type definitions. These parameters are stable no matter which directory
+// is getting generated.
+type ioContext struct {
+	mod             *modContext
+	input           bool
+	imports         map[string]codegen.StringSet
+	externalImports codegen.StringSet
+}
+
+// intoIOFiles converts this namespace into one or more files.
+// It recursively builds one file for each node in the tree.
+// If input=true, then it builds input types. Otherwise, it
+// builds output types.
+// The parameters in ctx are stable regardless of the depth of recursion,
+// but parent is expected to change with each recursive call.
+func (ns *namespace) intoIOFiles(ctx *ioContext, parent string) ([]*ioFile, error) {
+	// We generate the input and output namespaces when there are enums, regardless of i
+	// they are empty.
+	if ns == nil {
+		panic("TODO: The caller needs to check if this is nil or not first.")
+	}
+	fmt.Println(ns.Debug())
+	// Declare a new file to store the contents exposed at this directory level.
+	var dirRoot = path.Join(parent, ns.name)
+	var filename string
+	if ctx.input {
+		filename = path.Join(dirRoot, "inputs.ts")
+	} else {
+		filename = path.Join(dirRoot, "outputs.ts")
+	}
+	fmt.Printf("Generating namespace into IO file: %s\n", filename)
+	var file = newIOFile(filename)
+	// We start every file with the header information.
+	ctx.mod.genHeader(file.writer(), ctx.mod.sdkImports(true, false), ctx.externalImports, ctx.imports)
+	// We want to organize the items in the source file by alphabetical order.
+	sort.Slice(ns.types, func(i, j int) bool {
+		return tokenToName(ns.types[i].Token) < tokenToName(ns.types[j].Token)
+	})
+	sort.Slice(ns.enums, func(i, j int) bool {
+		return tokenToName(ns.enums[i].Token) < tokenToName(ns.enums[j].Token)
+	})
+	// Now, we write out the types declared at this directory
+	// level to the file.
+	var files = []*ioFile{file}
+	for i, t := range ns.types {
+		var isInputType = ctx.input && ctx.mod.details(t).inputType
+		var isOutputType = !ctx.input && ctx.mod.details(t).outputType
+		// Only write input and output types.
+		if isInputType || isOutputType {
+			if err := ctx.mod.genType(file.writer(), t, ctx.input, 0); err != nil {
+				return files, err
+			}
+			if i != len(ns.types)-1 {
+				fmt.Fprintf(file.writer(), "\n")
+			}
+		}
+	}
+	// We have successfully written all types at this level.
+	// Next, we recurse to generate the files at the next directory level.
+	sort.Slice(ns.children, func(i, j int) bool {
+		return ns.children[i].name < ns.children[j].name
+	})
+	for i, child := range ns.children {
+		// At this level, we export any nested definitions from
+		// the next level.
+		fmt.Fprintf(file.writer(), "import * as childname(%s) from \"childname(%s)\";\n", child.name, child.name)
+		fmt.Fprintf(file.writer(), "export type child(%s);", child.name)
```

Comment:
These are the re-exports? Interesting. 

I'm a bit confused how the childname and child() render.

Including the results of applying this codegen change to examples (PULUMI_ACCEPT=1 go test -test.v TestGeneratePackage) in the PR can help illusrate.

Also I would be vary of `export type`.  Syntactically consider re-exports https://www.typescriptlang.org/docs/handbook/modules.html#re-exports

Semantically consider the multiple levels here. `export type` will only export Type-level bindings. However I believe we might have value-level bindings in there as well? For example a class declaration has Type-level and Value-level bindings. There are also Namespace-level bindings. It's confusing.

I've just browsed through aws and tls types/{input,output}.ts. It would seem that most of the exports are exporting interfaces so `export type` is fine. There's a few places that do `export namespace`.

Reply:
Sorry, I tried to indicate previously (but it's been awhile and I should have written it down in line!) that this is the major WIP part of the PR. It's just absolutely not the right path. `childname` is a placeholder I added to help me trace the code.

> I've just browsed through aws and tls types/{input,output}.ts. It would seem that most of the exports are exporting interfaces so export type is fine. There's a few places that do export namespace.

I think each `export namespace` translates 1:1 with a new file created as part of this PR. When we create a new `namespace`, we're really trying to create a new layer of nesting. See the images attached below for how this translates.

Forgive me if I'm not explaining this well, I can my articulation is suffering from a lack of clarity.

![directories](https://user-images.githubusercontent.com/4955075/194125648-31e01042-5fd1-4126-8ff9-d17f9ec6cc12.svg)
![namespaces](https://user-images.githubusercontent.com/4955075/194125649-dc883eff-1618-4819-9aaa-06976762f5bd.svg)

Reply:
Trying to take your suggestion to use the `export * as x from "x"` syntax (from your linked resource) because you're right about the value-bindings. Right now it's looking like value bindings are still missing from those exports, which is strange, since they shouldn't be.